File: benchmark/runners/python/infermap_bench/compare.py
"""Compare a current report against a baseline, producing deltas."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


_METRIC_KEYS = ("f1", "top1", "mrr", "ece")
_ZERO_METRIC_SET = {k: 0.0 for k in _METRIC_KEYS} | {"n": 0}


@dataclass(frozen=True)
class Delta:
    """A computed delta between two reports (immutable)."""
    overall: dict[str, float] = field(default_factory=dict)
    by_difficulty: dict[str, dict[str, float]] = field(default_factory=dict)
    by_category: dict[str, dict[str, float]] = field(default_factory=dict)
    by_tag: dict[str, dict[str, float]] = field(default_factory=dict)
    per_case_deltas: list[tuple[str, float, float]] = field(default_factory=list)

    def is_regression(self, threshold: float = 0.02) -> bool:
        """True iff overall F1 dropped by *strictly* more than `threshold`.

        Uses a `1e-9` IEEE-754 epsilon guard so a drop exactly equal to
        `threshold` (modulo float rounding) is NOT classified as a regression.
        Mirrors the JS compare.ts `isRegression` implementation.
        """
        f1_delta = self.overall.get("f1", 0.0)
        return f1_delta < -threshold - 1e-9

    def top_movers(
        self,
        n: int = 10,
        threshold: float = 0.05,
    ) -> tuple[
        list[tuple[str, float, float, float]],
        list[tuple[str, float, float, float]],
    ]:
        """Return (regressions, improvements) — top-N cases by |Δf1|."""
        scored = [
            (cid, base, curr, curr - base)
            for cid, base, curr in self.per_case_deltas
        ]
        regressions = sorted(
            [d for d in scored if d[3] <= -threshold],
            key=lambda d: d[3],
        )[:n]
        improvements = sorted(
            [d for d in scored if d[3] >= threshold],
            key=lambda d: -d[3],
        )[:n]
        return (regressions, improvements)


def _metric_delta(baseline: dict[str, Any], current: dict[str, Any]) -> dict[str, float]:
    return {
        k: float(current.get(k, 0.0)) - float(baseline.get(k, 0.0))
        for k in _METRIC_KEYS
    }


def _slice_delta(
    baseline_slice: dict[str, Any],
    current_slice: dict[str, Any],
) -> dict[str, dict[str, float]]:
    out: dict[str, dict[str, float]] = {}
    keys = set(baseline_slice) | set(current_slice)
    for k in sorted(keys):
        b = baseline_slice.get(k, _ZERO_METRIC_SET)
        c = current_slice.get(k, _ZERO_METRIC_SET)
        out[k] = _metric_delta(b, c)
    return out


def compute_delta(baseline: dict, current: dict) -> Delta:
    """Diff two report.json dicts, producing a structured Delta.

    Computes per-metric deltas (`current - baseline`) for the overall scorecard
    and every by_* slice, plus per-case F1 triples for cases present in both
    reports. The returned Delta is frozen; all fields are populated at
    construction time.
    """
    baseline_cases = {c["id"]: c for c in baseline.get("per_case", [])}
    current_cases = {c["id"]: c for c in current.get("per_case", [])}
    common = sorted(set(baseline_cases) & set(current_cases))
    return Delta(
        overall=_metric_delta(
            baseline["scorecard"]["overall"], current["scorecard"]["overall"]
        ),
        by_difficulty=_slice_delta(
            baseline["scorecard"].get("by_difficulty", {}),
            current["scorecard"].get("by_difficulty", {}),
        ),
        by_category=_slice_delta(
            baseline["scorecard"].get("by_category", {}),
            current["scorecard"].get("by_category", {}),
        ),
        by_tag=_slice_delta(
            baseline["scorecard"].get("by_tag", {}),
            current["scorecard"].get("by_tag", {}),
        ),
        per_case_deltas=[
            (cid, float(baseline_cases[cid]["f1"]), float(current_cases[cid]["f1"]))
            for cid in common
        ],
    )

File: benchmark/runners/ts/src/compare.ts
// Compare a current report against a baseline, producing deltas.
// Mirrors infermap_bench/compare.py.

interface MetricSet {
  f1: number;
  top1: number;
  mrr: number;
  ece: number;
  n: number;
}

interface Report {
  version: 1;
  language: string;
  infermap_version: string;
  runner_version: string;
  ran_at: string;
  duration_seconds: number;
  scorecard: {
    overall: MetricSet;
    by_difficulty: Record<string, MetricSet>;
    by_category: Record<string, MetricSet>;
    by_tag: Record<string, MetricSet>;
  };
  per_case: Array<{
    id: string;
    f1: number;
    top1: number;
    mrr: number;
    expected_n: number;
    predicted_n: number;
    true_positives: number;
    false_positives: number;
    false_negatives: number;
    failed: boolean;
    failure_reason: string | null;
  }>;
  failed_cases: string[];
}

const METRIC_KEYS = ["f1", "top1", "mrr", "ece"] as const;
type MetricKey = (typeof METRIC_KEYS)[number];

const ZERO_METRIC_SET: MetricSet = { f1: 0, top1: 0, mrr: 0, ece: 0, n: 0 };

export interface Mover {
  caseId: string;
  baselineF1: number;
  currentF1: number;
  deltaF1: number;
}

export class Delta {
  constructor(
    public readonly overall: Record<MetricKey, number> = { f1: 0, top1: 0, mrr: 0, ece: 0 },
    public readonly byDifficulty: Record<string, Record<MetricKey, number>> = {},
    public readonly byCategory: Record<string, Record<MetricKey, number>> = {},
    public readonly byTag: Record<string, Record<MetricKey, number>> = {},
    public readonly perCaseDeltas: ReadonlyArray<readonly [string, number, number]> = [],
  ) {}

  /**
   * True iff overall F1 dropped by strictly more than `threshold`.
   * Uses a `1e-9` IEEE-754 epsilon guard so a drop exactly equal to
   * `threshold` is NOT classified as a regression.
   */
  isRegression(threshold = 0.02): boolean {
    const f1Delta = this.overall.f1 ?? 0;
    return f1Delta < -threshold - 1e-9;
  }

  /** Return top-N regressions and improvements by |ΔF1|. */
  topMovers(n = 10, threshold = 0.05): { regressions: Mover[]; improvements: Mover[] } {
    const scored: Mover[] = this.perCaseDeltas.map(([caseId, baselineF1, currentF1]) => ({
      caseId,
      baselineF1,
      currentF1,
      deltaF1: currentF1 - baselineF1,
    }));
    const regressions = scored
      .filter((m) => m.deltaF1 <= -threshold)
      .sort((a, b) => a.deltaF1 - b.deltaF1)
      .slice(0, n);
    const improvements = scored
      .filter((m) => m.deltaF1 >= threshold)
      .sort((a, b) => b.deltaF1 - a.deltaF1)
      .slice(0, n);
    return { regressions, improvements };
  }
}

function metricDelta(baseline: MetricSet, current: MetricSet): Record<MetricKey, number> {
  const out = {} as Record<MetricKey, number>;
  for (const k of METRIC_KEYS) out[k] = current[k] - baseline[k];
  return out;
}

function sliceDelta(
  baseline: Record<string, MetricSet>,
  current: Record<string, MetricSet>,
): Record<string, Record<MetricKey, number>> {
  const keys = new Set([...Object.keys(baseline), ...Object.keys(current)]);
  const sorted = [...keys].sort();
  const out: Record<string, Record<MetricKey, number>> = {};
  for (const k of sorted) {
    out[k] = metricDelta(baseline[k] ?? ZERO_METRIC_SET, current[k] ?? ZERO_METRIC_SET);
  }
  return out;
}

export function computeDelta(baseline: Report, current: Report): Delta {
  const overall = metricDelta(baseline.scorecard.overall, current.scorecard.overall);
  const byDifficulty = sliceDelta(
    baseline.scorecard.by_difficulty ?? {},
    current.scorecard.by_difficulty ?? {},
  );
  const byCategory = sliceDelta(
    baseline.scorecard.by_category ?? {},
    current.scorecard.by_category ?? {},
  );
  const byTag = sliceDelta(
    baseline.scorecard.by_tag ?? {},
    current.scorecard.by_tag ?? {},
  );

  const baselineCases = new Map<string, number>();
  for (const c of baseline.per_case ?? []) baselineCases.set(c.id, c.f1);
  const currentCases = new Map<string, number>();
  for (const c of current.per_case ?? []) currentCases.set(c.id, c.f1);

  const common: string[] = [];
  for (const id of baselineCases.keys()) {
    if (currentCases.has(id)) common.push(id);
  }
  common.sort();
  const perCaseDeltas: Array<readonly [string, number, number]> = common.map(
    (id) => [id, baselineCases.get(id)!, currentCases.get(id)!] as const,
  );

  return new Delta(overall, byDifficulty, byCategory, byTag, perCaseDeltas);
}
